python/kfserving/kfserver/storage.py

```diff
@@ -0,0 +1,44 @@
+import os
+import shutil
+import errno
+import logging
+
+_GCS_PREFIX = "gs://"
+_S3_PREFIX = "s3://"
+_LOCAL_PREFIX = "/"
+
+
+class Storage(object):
```

Comment:
wonder if its possible to inject an init container to download stuff.. will allow this to be independent of storage implementations and will also potentially enable caching... just a thought possibly for a later date?

Reply:
Caching with Persistent Volumes?

Reply:
yes - we have some teams that have implemented this on k8s.. but the download should be decoupled from the caching obviously. In general will keep the container small and avoid having it implement download from all clouds (and keep up with versions)...

Reply:
So for a side-car are we saying the servers always expect their models to be in a hardwired local location? As long as easy standalone testing is possible without the sidecar. Certainly all the issues with Authentication can then be handled by this separate piece of logic which might be good?

Reply:
Here's an alternative thought. What if we used a knative build to bundle the model into the serving container?

Kubernetes (GKE especially) has lots of optimizations for image caching. We effectively get PVs for free. It also achieves decoupling the storage auth from the service.

Reply:
this thought seems at odds with the server multi-tenancy we discussed?

Reply:
Agreed, but consider baking n models. It really depends on how dynamic we want the multi-tenancy loading/unloading to be. I worry that the eventual consistency of DNS-based routing will make it challenging for us to do too much dynamic loading/unloading. The problem with baking is that we may end up with combinatorial images (sparcely built) or huge images (densly built).

We can always support a lower cold start for single models, and the multi-model-server might suffer some tradeoffs.